include/circt/Dialect/HW/HWModuleGraph.h

```diff
@@ -0,0 +1,189 @@
+    : public llvm::DefaultDOTGraphTraits {
+  using DefaultDOTGraphTraits::DefaultDOTGraphTraits;
+
+  static std::string getNodeLabel(circt::hw::detail::HWOperation *node,
+                                  circt::hw::HWModuleOp) {
+    return llvm::TypeSwitch<mlir::Operation *, std::string>(node)
+        .Case<circt::comb::AddOp>([&](auto) { return "+"; })
+        .Case<circt::comb::SubOp>([&](auto) { return "-"; })
+        .Case<circt::comb::AndOp>([&](auto) { return "&"; })
+        .Case<circt::comb::OrOp>([&](auto) { return "|"; })
+        .Case<circt::comb::XorOp>([&](auto) { return "^"; })
+        .Case<circt::comb::MulOp>([&](auto) { return "*"; })
+        .Case<circt::comb::MuxOp>([&](auto) { return "mux"; })
+        .Case<circt::comb::ShrSOp, circt::comb::ShrUOp>(
+            [&](auto) { return ">>"; })
+        .Case<circt::comb::ShlOp>([&](auto) { return "<<"; })
+        .Case<circt::comb::ICmpOp>([&](auto op) {
+          switch (op.getPredicate()) {
+          case circt::comb::ICmpPredicate::eq:
+          case circt::comb::ICmpPredicate::ceq:
+          case circt::comb::ICmpPredicate::weq:
+            return "==";
+          case circt::comb::ICmpPredicate::wne:
+          case circt::comb::ICmpPredicate::cne:
+          case circt::comb::ICmpPredicate::ne:
+            return "!=";
+          case circt::comb::ICmpPredicate::uge:
+          case circt::comb::ICmpPredicate::sge:
+            return ">=";
+          case circt::comb::ICmpPredicate::ugt:
+          case circt::comb::ICmpPredicate::sgt:
+            return ">";
+          case circt::comb::ICmpPredicate::ule:
+          case circt::comb::ICmpPredicate::sle:
+            return "<=";
+          case circt::comb::ICmpPredicate::ult:
+          case circt::comb::ICmpPredicate::slt:
+            return "<";
+          }
+          assert(false && "unhandled ICmp predicate");
+        })
+        .Case<circt::seq::CompRegOp>([&](auto) { return "reg"; })
+        .Case<circt::hw::ConstantOp>([&](auto op) {
+          llvm::SmallString<64> valueString;
+          op.getValue().toString(valueString, 10, false);
+          return valueString.str().str();
+        })
+        .Default([&](auto op) { return op->getName().getStringRef().str(); });
+  }
+
+  std::string getNodeAttributes(circt::hw::detail::HWOperation *node,
+                                circt::hw::HWModuleOp) {
+    std::string attrs;
+
+    llvm::TypeSwitch<mlir::Operation *>(node)
+        .Case<circt::hw::ConstantOp>(
+            [&](auto) { attrs = "fillcolor=darkgoldenrod1,style=filled"; })
+        .Case<circt::comb::MuxOp>([&](auto) {
+          attrs = "shape=invtrapezium,fillcolor=bisque,style=filled";
+        })
+        .Case<circt::hw::OutputOp>(
+            [&](auto) { attrs = "fillcolor=lightblue,style=filled"; })
+        .Default([&](auto op) {
+          llvm::TypeSwitch<mlir::Dialect *>(op->getDialect())
+              .Case<circt::comb::CombDialect>([&](auto) {
```

Comment:
Should you also handle the hw dialect for data structure stuff? Or do you like the defaults for that?

Reply:
In this commit, i'd prefer sticking to defaults. As mentioned in the commit message, **a lot** of additional stuff could be done to pretty up this graph. My hunch here is that if we'd want to have graphs that are "presentable" and not just a small debugging tool, then `GraphTrait`s should be ditched in favor of manually printing a graph (allowing us to have much better control over clusters, ids, labels, ...).